### ISSUE TITLE
`<Form>`

### ISSUE THREAD

**Issue by Rich-Harris** [sveltejs/kit#3533]
### Describe the problem

Best practices around form data are slightly annoying — we want to encourage the use of native `<form>` behaviour as far as possible (accessible, works without JS, etc) but the best UX involves optimistic updates, pending states, and client-controlled navigation.

So far our best attempt at squaring this circle is the `enhance` action [found in the project template](https://github.com/sveltejs/kit/blob/f2d64256aecc7419ed6a8d424ef00686dbf8b33a/packages/create-svelte/templates/default/src/lib/form.ts). It's a decent start, but aside from being tucked away in an example rather than being a documented part of the framework itself, actions are fundamentally limited in what they can do as they cannot affect SSR (which means that method overriding and CSRF protection will always be relegated to userland).

Ideally we would have a solution that

1. was a first-class part of the framework
2. enabled best practices _and_ best UX
3. worked with SSR

### Describe the proposed solution

I propose adding a `<Form>` component. By default it would work just like a regular `<form>`...

```svelte
<script>
  import { Form } from '$app/navigation';
</script>

<Form action="/todos.json" method="post">
  <input name="description" aria-label="Add todo" placeholder="+ tap to add a todo">
</Form>
```

...but with additional features:

### Automatic invalidation

Using the [`invalidate`](https://kit.svelte.dev/docs#modules-$app-navigation) API, the form could automatically update the UI upon a successful response. In the example above, using the form with JS disabled would show the endpoint response, meaning the endpoint would typically do something like this:

```js
export async function post({ request, locals }) {
  const data = await request.formData();

  // write the todo to our database...
  await db.write('todos', locals.user, data.get('description'));

  // ...then redirect the user back to /todos so they see the updated page
  return {
    status: 303,
    headers: {
      location: '/todos'
    }
  };
}
```

If JS isn't disabled, `<Form>` would submit the result via `fetch`, meaning there's no need to redirect back to the page we're currently on. But we _do_ want the page to reflect the new data. Assuming (reasonably) that the page is showing the result of a `GET` request to `/todos.json`, `<Form>` can do this automatically:

```js
// (this is pseudo-code, glossing over some details)

const url = new URL(action, location);
url.searchParams.set(method_override.parameter, method);

const res = await fetch(url, {
  method: 'post',
  body: new FormData(form)
});

// if the request succeeded, we invalidate the URL, causing
// the page's `load` function to run again, updating the UI
if (res.ok) {
  invalidate(action);
}
```

### Optimistic UI/pending states

For some updates it's reasonable to wait for confirmation from the server. For others, it might be better to immediately update the UI, possibly with a pending state of some kind:

```svelte
<Form
  action="/todos.json"
  method="post"
  pending={({ data }) => {
    // we add a new todo object immediately — it will be destroyed
    // by one without `pending: true` as soon as the action is invalidated
    todos = [...todos, {
      done: false,
      text: data.get('text'),
      pending: true
    }];
  }}
  done={({ form }) => {
    // clear the form
    form.reset();
  }}
>
  <input name="description" aria-label="Add todo" placeholder="+ tap to add a todo">
</Form>

{#each todos as todo}
  <div class="todo" class:done={todo.done} class:pending={todo.pending}>
    <!-- todo goes here --> 
  </div>
{/each}

<style>
  .pending {
    opacity: 0.5;
  }
</style>
```

> this example glosses over one problem — typically you might generate a UUID on the server and use that in a keyed each block or something. ideally the pending todo would also have a UUID that the server accepted so that the invalidation didn't cause glitches. need to think on that some more

### Error handling

There's a few things that could go wrong when submitting a form — network error, 4xx error (e.g. invalid data) or 5xx error (the server blew up). These are currently handled a bit inconsistently. If the handler returns an explicit error code, the page just shows the returned `body`, whereas if an error is thrown, SvelteKit renders the error page. 

#3532 suggests a way we can improve error handling, by rendering a page with validation errors. For progressively enhanced submissions, this wouldn't quite work — invalidating the action would cause a `GET` request to happen, leaving the validation errors in limbo. We can pass them to an event handler easily enough...

```svelte
<Form
  action="/todos.json"
  method="post"
  pending={({ data }) => {...}}
  done={({ form }) => {...}}
  error={async ({ response }) => {
    ({ errors } = await response.json());
  }}
/>
```

...but we don't have a great way to _enforce_ correct error handling. Maybe we don't need to, as long as we provide the tools? Need to think on this some more.

### Method overriding

Since the component would have access to the [`methodOverride`](https://kit.svelte.dev/docs#routing-endpoints-http-method-overrides) config, it could override the method or error when a disallowed method is used:

```svelte
<Form action="/todos/{todo.uid}.json" method="patch">
  <input aria-label="Edit todo" name="text" value={todo.text} />
  <button class="save" aria-label="Save todo" />
</Form>
```

### CSRF

We still need to [draw the rest of the owl](#72):

<img width="927" alt="image" src="https://user-images.githubusercontent.com/1162160/150874245-95a2fa3b-3a1f-4748-85cb-493bc94d3079.png">

I think `<Form>` has an important role to play here though. It could integrate with Kit's hypothetical CSRF config and automatically add a hidden input...

```svelte
<!-- <Form> implementation -->
<script>
  import { csrf } from '$app/env'; // or somewhere
</script>

<form {action} {method} on:submit|preventDefault={handler}>
  <input type="hidden" name={csrf.key} value={csrf.value}>
  <slot/>
</form>
```

...which we could then _somehow_ validate on the server. For example — this may be a bit magical, but bear with me — maybe we could intercept `request.formData()` and throw an error if CSRF checking (most likely using the double submit technique) fails? We could add some logic to our [existing response proxy](https://github.com/sveltejs/kit/blob/8a5f5de72c7b55a3ce230c50ac904b291096da77/packages/kit/src/runtime/server/page/load_node.js#L208):

```js
// pseudo-code
const proxy = new Proxy(response, {
  get(response, key, _receiver) {
    if (key === 'formData') {
      const data = await response.formData();
      const cookies = cookie.parse(request.headers.get('cookie'));

      // check that the CSRF token the page was rendered with 
      // matches the cookie that was set alongside the page
      if (data.get(csrf.key) !== cookies[csrf.cookie]) {
        throw new Error('CSRF checking failed');
      }

      return data;
    }
  }

  // ...
});
```

This would protect a lot of users against CSRF attacks without app developers really needing to do anything at all. We _would_ need to discourage people from using `<Form>` on prerendered pages, of course, which is easy to do during SSR.

### Alternatives considered

The alternative is to leave it to userland. I don't _think_ I've presented anything here that requires hooks into the framework proper — everything is using public APIs (real or hypothetical). But I think there's a ton of value in having this be built in, so that using progressively-enhanced form submissions is seen as the right way to handle data.

This is a big proposal with a lot of moving parts, so there are probably a variety of things I haven't considered. Eager to hear people's thoughts.

### Importance

would make my life easier

**Comment by kevmodrome** [comment]
I generally like this idea. I've been using the form action and it's very nice.

How do we handle styling? Using an action it's easy since you just leave it to the user, with a component you need some kind of API.

**Comment by dummdidumm** [comment]
I like this idea as well. Some things that are not clear yet, they are essentially about how far does "progressive enhancement" go. For example, let's say I want to provide better DX and do client-side validation (like: on blur the field is marked as invalid because it's required and there's no content in it). How to achieve this? Another use case: You have a component library which wraps HTML input elements of different kinds - is it still possible to interact with them in an easy way? An advanced use case: You have a list of form items and you can add more items by clicking a `+` button - will the form be able to detect this dynamic addition of another form element?

These are all use cases you don't need for a simple sign up form but which will come up in internal business apps quite often.

**Comment by kevmodrome** [comment]
> I like this idea as well. Some things that are not clear yet, they are essentially about how far does "progressive enhancement" go. For example, let's say I want to provide better DX and do client-side validation (like: on blur the field is marked as invalid because it's required and there's no content in it). How to achieve this? Another use case: You have a component library which wraps HTML input elements of different kinds - is it still possible to interact with them in an easy way? An advanced use case: You have a list of form items and you can add more items by clicking a `+` button - will the form be able to detect this dynamic addition of another form element?

I think the correct way to handle form validation is to rely heavily on the platform - nudge users in the direction of using CSS.  The MDN article here is very thorough: https://developer.mozilla.org/en-US/docs/Learn/Forms/Form_validation. The advantage of this is that it would work both in js and non-js situations.

[Here's a very barebones example](https://svelte.dev/repl/e331da83fd8d43ce99df70242f6cc47b?version=3.46.2) of we might encourage users to do.

**Comment by Rich-Harris** [comment]
> How do we handle styling?

Styles that you would have applied to the form itself probably need to go on a wrapper element either inside or outside the component. I think this is a reasonably small price to pay

> For example, let's say I want to provide better DX and do client-side validation (like: on blur the field is marked as invalid because it's required and there's no content in it). How to achieve this?

We could add a presubmit callback with the ability to abort, but honestly most of the things that it's possible to validate client-side (ie not stuff like 'username already taken') can be done with HTML attributes, and we should push people in that 

> You have a component library which wraps HTML input elements of different kinds - is it still possible to interact with them in an easy way? An advanced use case: You have a list of form items and you can add more items by clicking a + button - will the form be able to detect this dynamic addition of another form element?

The component doesn't know what it contains and it doesn't need to know - `new FormData(form)` only cares what's in the DOM at submit time

**Comment by f-elix** [comment]
I'm so glad this is brought up! I have really dug into the Remix `<Form>` component recently and implemented a version of it in Sveltekit (not a package, at least yet). As things are right now I have ro resort to some hacks (like passing action data to `event.locals` and populating the `session` with it to retrieve it in my components), so I would absolutely love if Sveltekit provided an official solution.

What you're proposing seems really sensible to me. 

For styling, I think you could either expose `class` and `style` props, or juste use `$$restProps` and allow any attribute so they can be used for styling (like data attributes and such). Maybe there are problems I'm not thinking of with such a blunt approach though.

One quirk I've run into is adding events dynamically on the `form` itself, like a `reset` event, or `input`/`change` events (so we can use event delegation and not add a listener on every field). Right now I'm forwarding the native events that I want to use because Svelte doesn't support dynamic events. That is of course more related to Svelte than Sveltekit, but I think its worth mentioning, since its a use case a lot of people might run into.

**Comment by arxpoetica** [comment]
> For styling, I think you could either expose class and style props

Duh. I thought of exposing `style` (and hated just that), but exposing both solves the styling problem. Do that @Rich-Harris and I'll stop complaining. 😉

**Comment by HibiscusCoffee** [comment]
I've been doing this with my own form component for a while and think it makes form handling a lot easier. I've chosen to handle validation with JS though so that the same ‘validate‘ function can be called programmatically on the client, on blur and on the server. Not 100% sure about this but it makes it possible to validate the response a second time on the server and if JS is disabled in the browser the error messages are simply included in the document, server rendered and send back to the client.

**Comment by oliie** [comment]
I also like this idea! An other idea is to maybe have it build in already instead of importing Form, and maybe use something like `<svelte:form>` instead?

**Comment by brittneypostma** [comment]
> > How do we handle styling?
> 
> Styles that you would have applied to the form itself probably need to go on a wrapper element either inside or outside the component. I think this is a reasonably small price to pay
> 

There could be props created on the form component to allow for custom styling also or passing them via style props.

**Comment by dangelomedinag** [comment]
I'm not entirely sure that a `<Form>` component is the right solution. The reasons for thinking this are the following:

- Svelte/Sveltekit does not provide components like other frameworks eg. `<Link>`, on the other hand, implements a solution directly on the HTML element. (and that's awesome).
- It is highly likely that the solution of creating a `<Form>` component will not fit the needs of a large project where you need to manipulate a lot of things, both the element and logic. (you will end up implementing your own solution).
- Svelte/Sveltekit it provides enough tools to do it on our own, without making us feel helpless by the framework itself.
- The infinite dilemma of how to apply styles to nested components comes into play. using an "action" maintains the existing advantages (and disadvantages) up to now in terms of styles.

it seems like a better idea to provide an "action" than a component. the difference in implementation will not be too different from one another.

`<Form>` component:
```svelte
<script>
  import { Form } from '$app/navigation';
</script>
<Form
  action="/api/test"
  method="PUT" // method overrides
  class="flex flex-column p-4 mt-2"
  id="FormComponent"
  on:pending={(e) => console.log(e.detail /* => data */)}
  on:result={(e) => console.log(e.detail /* => response */)}
  on:error={(e) => console.log(e.detail /* => Error handler */)}
>
```

`<form>` with action
```svelte
<script>
  import { formEnhance } from '$app/navigation';
</script>
<form
  action="/api/test?_method=PUT" // manually method overrides
  method="post"
  class="flex flex-column p-4 mt-2"
  id="nativeFormELement"
  use:formEnhance={{
    pending: (data) => console.log(data),
    result: async (res) => console.log(await res.json()),
    error: (e) => console.log(e.message)
  }}
>
```

PS: my language is Spanish, the text is translated, sorry if there is something strange.

**Comment by PierBover** [comment]
IMO the big advantage of using an action like `use:form` is that it returns control to of the whole form to the parent component.

No more styling issues but also, what if we want to use custom actions for form validation and whatnot? Or what if we need to get a ref to form DOM element?

**Comment by lukaszpolowczyk** [comment]
I'm against locking such things into a component.
It's not a universal solution.

But if you want the benefits of a component, you might consider implementing the Declarative Actions proposal:
https://github.com/sveltejs/rfcs/pull/41

Then you have the benefits of the component, and you don't lose out on the drawbacks of the component, because you still have the full element exposed.

There's magic going on underneath, and it's regular elements on top.

**Comment by lukaszpolowczyk** [comment]
@dangelomedinag Moreover, through `use:action` you can create an event listener, then it will look like this:
```svelte


<script>
  import { formEnhance } from '$app/navigation';
</script>
<form
  action="/api/test?_method=PUT" // manually method overrides
  method="post"
  class="flex flex-column p-4 mt-2"
  id="nativeFormELement"
  on:pending={(e) => console.log(e.detail /* => data */)}
  on:result={(e) => console.log(e.detail /* => response */)}
  on:error={(e) => console.log(e.detail /* => Error handler */)}
  use:formEnhance
>
```
And through [Declarative Actions](https://github.com/sveltejs/rfcs/pull/41) it will be even easier to design.

**Comment by Glench** [comment]
Really glad you're thinking about this. In my own SvelteKit projects I've found it very tedious to implement all the form plumbing, and doing it consistently has proven difficult with the result that each page in my app has different ways of doing things.

So it would be a great relief to have one consistent, framework-provided, batteries included solution that is a sensible default, but of course be able to fall back if needed.

#### Pending, pre-submit, post-submit

As for pending updates, pre-submit, and post-submit, one simple solution I've found that handles all of these cases in one is to provide the form's special "do the network request submission stuff" in a function that the user can call:

```svelte
<script>
function submit({ network_request_magic }) {
  // do anything you want before data is sent to the server
  const res = await network_request_magic(); // network request stuff happens here
  // do anything you want after response is received from the server
}
</script>

<Form action="" on:submit={submit}>
...
</Form>
```

One problem with this approach is that if the user potentially leaves out a call to `network_request_magic` then the form may not work.

#### Slot prop

Also, if there is indeed a `<Form>` component, it would be great to expose the pending network status as a slot prop:

```svelte
<Form let:loading={loading}>
  {#if loading}
    <SpinnerThingy />
  {/if}
  <input disabled={loading} />
</Form>
```

I do this in my own Form component and it's great — don't even need to have a `<script>` tag.

**Comment by cdcarson** [comment]
My take: The javascript-disabled user is an edge case. That doesn't mean that it shouldn't be addressed, just that adressing it should not be central to the SvelteKit API. The approach should be as agnostic and as standard as possible. Ideally, there would be no pre-baked "SvelteKit way".

SvelteKit already handles plain HTML forms just fine. What's needed is documentation and example code that show how to do it. Here are some things that might be addressed:

- Persisting form state from a `POST` across a redirect to a page. 
- Making client and endpoint code agnostic as to XHR (fetch) vs plain HTML.
- CSRF.


Of these, persisting state is the 900 lb gorilla, the one I suspect is driving the discussion here about `<Form/>` and [on this thread about shadow endpoints](https://github.com/sveltejs/kit/issues/3532#:~:text=easily%20be%20enforced.-,POST%20requests,-This%20becomes%20even). Maybe I have the wrong end of the stick, but the notion seems to be that a `POST` endpoint would both handle the form and automatically return the HTML page, nicely populated with the form state, all in one go. 

I understand the attraction of this idea. You don't need to persist & redirect. But in my experience mixing verbs always ends up being fragile, inconvenient, and hard to reason about. It's much better in the long run to persist state and redirect, as complex or unnecessary as that seems in the context of one or two routes. 

Net. Dealing with form state and the other concerns relevant to the  case is really not that hard and need not be abstracted away by SvelteKit. I'd argue that abstractions like `<Form/>` and conventional magic like "shadow endpoints" may pose more of a cognitive hurdle than, say, dealing with Redis or understanding CSRF protection. Further, I'd worry that things end up driving the shape of the overall framework API in service of (what I think should be) an implementation detail best left to userland.

Anyway, if you think the documentation/example route is the way to go I'd be happy to help out.

**Comment by Rich-Harris** [comment]
Re using an action rather than a component — I do understand the appeal, but it does mean adding a ton of boilerplate for CSRF in particular. One appeal of `<Form>` is that we could pretty much protect users of SvelteKit apps against CSRF attacks without app developers even needing to know what that means.

**Comment by Tropix126** [comment]
Also: what should be forwarded to `<Form />`? Should it use $$restProps? What events should be available? There are ways around things like `bind:this` (i.e. exporting an `element` prop that is bound internally, then using `bind:element` in parent context), but I imagine something like that would go against what people are typically used to. The limitations of components are fairly high atm.

**Comment by lukaszpolowczyk** [comment]
@Rich-Harris See Declarative Actions:
https://github.com/sveltejs/rfcs/pull/41

With **Declarative Actions** you can get **all the benefits of the Component**, but there would be no hidden properties of the HTML element.

Because the **problem with `use:action`** in this case, and in general, is that it is a **simple function, not using any of the Component's benefits**.

If use:form would be Declarative Action, then the problem disappears.

Declarative Actions is @Zizico2's proposal and here I just (more or less) wrote how a regular Component compares to a Declarative Action:
https://github.com/sveltejs/rfcs/pull/41#issuecomment-934364393
In short: More convenient to use HTML element attributes, need to dynamically pass special attributes to the element inside the component(in my example `$$specialProps`, but there is also a [forward directive proposal](https://github.com/sveltejs/rfcs/pull/60) from @Tropix126).

When done this way, Declarative Action will be like a regular component, but with enhancements and without hiding the element attributes.
So **SSR etc. will be doable with Declarative Action**, so **I guess CSRF will be just as easy to deal with too?**

**Comment by arxpoetica** [comment]
@Rich-Harris 
> Re using an action rather than a component — I do understand the appeal, but it does mean adding a ton of boilerplate for CSRF in particular.

Can you illustrate what you mean in terms of code? What might that boilerplate look like?

I admit I'm pretty excited by the idea of actions/`use:` instead of locking away the component.

**Comment by PierBover** [comment]
> Btw, this doesn't work if you remove bind:value. Why?

Because `input.validity` is not reactive.

I'm guessing when `value` updates, it forces a render of the if block.

**Comment by kevmodrome** [comment]
Here's an old library I made once upon a time. It has an action that serialises and deserialises the form data (based on lukeeds excellent library btw!): https://github.com/svelteschool/svelte-forms/blob/master/src/actions/form/getValues.js

**Comment by f-elix** [comment]
> A better approach is to use an action on the <form> element which would allow to get the events when these bubble up, or to add event listeners to all inputs (automatically from the action). Unfortunately, with the <Form> component Rich is proposing here, I don't know how that would work since actions cannot be used on components.

You could also add an `input` or `change` listener (or both) on the form and use event delegation to get the values.

**Comment by PierBover** [comment]
> Here's an old library I made once upon a time. It has an action that serialises and deserialises the form data (based on lukeeds excellent library btw!): https://github.com/svelteschool/svelte-forms/blob/master/src/actions/form/getValues.js

Cool trick dispatching the custom event!

**Comment by timwis** [comment]
Love the idea of CSRF protection being automatic and default, like in rails.

Can I suggest, though, that it migth be worth giving it a slightly different name than `<Form>`, if only for readability's sake?

```html
<Form action="/todos.json" method="post">
  <input name="description" aria-label="Add todo" placeholder="+ tap to add a todo">
</Form>
```

Scanning through a file visually, it's a rather subtle thing to notice that isn't a regular `<form>`.

Alternatives might be `<svelte:form>` as @oliie suggested, which would be consistent with the other 7 built-in components, or a namespace, e.g. `<Svelte.Form>`, or even a two-word `<SvelteForm>` would be a lot easier to notice.

**Comment by Mlocik97** [comment]
Now that I think about forms, I think we don't need `<Form>` resp. `<svelte:form>` component, as crsf is solvable with just standard HTML forms, and hydratation is solved by Svelte action, so why don't we just that action that is in demo app, make it importable from kit? Also I think it would make code more natural and simpler.

**Comment by mattcroat** [comment]
I welcome a change like this because I believe most people are going to only use JavaScript if it's not easy!

As a content creator that made a series on SvelteKit it was weird having to explain progressive enhancement but leaving that responsibility on the user. The original example is hidden but what makes it worse are the types that make it unreadable.

The SvelteKit documentation doesn't have a single example of using forms.

Not only the documentation but during my research I could barely find any examples in videos or posts I consumed that take advantage of it because it's not something that's encouraged from the documentation.

**Comment by ITwrx** [comment]
A form component might also be an opportunity to remove more form and error handling boilerplate from the `.svelte` and `.js/.ts` endpoint files.  Maybe it could transparently invalidate a `Store` associated with a form when appropriate too? I haven't used `Stores` yet, so maybe that is ill-conceived.

**Comment by tsongas** [comment]
Forms are a major pain, they can easily take up 50% or more of the effort of building an application. The amount of code needed to wire up something as simple as a list of dynamically generated checkboxes using libraries like Formik and react-hook-form is staggering. Doing something to improve this situation could be Svelte and SvelteKit's single biggest win for developers.

**Comment by Mlocik97** [comment]
I don't know these libs, but implementing my own solution with forms is really really easy with enchance action.

**Comment by groovy9** [comment]
I'm and end-user type of web developer; the browser and framework internals are above my pay grade, but I make business management apps.  My app suite is in need of a rewrite and I'm going with sveltekit and I was really excited about the endpoint feature.  

Most of what I do starts with grabbing a data table from a database with some sane defaults (like only active employees or only jobs scheduled for today), filtering it down and then viewing/editing a record.  Basic business app type stuff.

I thought surely if there's a feature like endpoints, it must also have simplified querying the same endpoint over and over without a page-refreshing form submit, but was disappointed that it hasn't.  I have it working with code like below, which is fine and not any worse than any other framework but all this is to say that I would just about kill for some tight integration between a component and its associated endpoint, such that I don't have to manually worry about building fetch requests.  

Here's a vote for a custom Form tag with a prop that tells it to hit the JSON endpoint again and replace the data without a page refresh.  If I'm overlooking a simple way to do this today, please clue me in.

```
// quick and dirty, don't judge me
<script lang="ts">
  import { browser } from "$app/env";
  export let rows: any[];
  let data: any[] = rows;
  let age: number = undefined;

  $: update(age);

  async function update(age: number) {
    if (browser) {
      let url = `/stuff/__data.json`;
      if (age) url += `?age=${age}`;
      const response = await fetch(url);
      let d = await response.json();
      data = d.rows;
    }
  }
</script>

<form>
  Age: <input type="text" name="age" bind:value={age} />
</form>
```

**Comment by Mlocik97** [comment]
You already can have:

```html
<script>
    import { enhance } from '$lib/form';
    export let data = { something: "" };

   function addData({data, form, response}) {
        if (response.ok) {
            data = await response.json();
        }
        form.reset();
   }
</script>

<form action="/endpoint" method="POST" use:enhance={{result: addData}}>
    <input type="text" value={data.something} name="something>
    <button type="submit" 
</form>
```

it works fine with and without JS, and you don't need reload when JS is enabled, it hydrates as you need....

in `$lib/form` you have: <https://github.com/sveltejs/kit/blob/master/packages/create-svelte/templates/default/src/lib/form.ts>

**Comment by groovy9** [comment]
@[Mlocik97](https://github.com/Mlocik97), that looked great on the surface, but I'm not sure having an imported custom enhance function that I'll have no memory of in 6 months is less trouble than just composing fetches right in the .svelte files.  And anyway, it seems to have some screwy behavior.  For example, if I do a form.submit() when a form field updates, it does a full page refresh whereas if I do a submitButton.click() it works as desired.  But both of those things should be equivalent, no?  

Anyway, the moral of the story for me that since this is such an insanely common use case, I'd like the framework to make it dead simple.  And sveltekit **almost** does with endpoints, but stops just a bit short.

**Comment by f-elix** [comment]
@groovy9 As an aside, `form.submit()` and `submitButton.click()` does not work the same. The full page refresh that occurs when you call `form.submit()` is the normal behavior of that method. 

You can try using `form.requestSubmit()` instead (https://developer.mozilla.org/en-US/docs/Web/API/HTMLFormElement/requestSubmit).

The enhance function works as intended.

**Comment by groovy9** [comment]
@f-elix ah, yep, that works.  And I guess the last mystery is that it always follows up the POST that requestSubmit() does with a GET to the same endpoint, the same one it does on the initial page load.  If I change the form action to some other endpoint, it just POSTs that endpoint like I'd expect without the extra GET.

```
// relevant parts of actual working code
<script lang="ts">
  export let rows: any[]; 
  let data: any[] = rows; 
  ...
  $: resubmit(age)

  function resubmit(age) {
    form?.requestSubmit()  // triggers POST plus extra GET to /endpoint/__data.json
  }

  async function addData({ response }) {
    if (response.ok) {
      let d = await response.json();  // only sees the POST results
      data = d.rows;
    }
  }
</script>

<form method="post" use:enhance={{ result: addData }} bind:this={form}>
  Age: <input type="text" name="age" bind:value={age} />
</form>

**Comment by mattcroat** [comment]
@groovy9 If you want you can use [Felte](https://felte.dev/) that does it for you but also adds more on top and makes working with forms easier until it's part of the framework.

**Comment by SrGeneroso** [comment]
I like svelte being opinionated and forms are a fundamental part of web development so I'm eager to see some brilliant svelte magic around this topic.
Is true HTML5 has some nice features by itself but it could be much better.

**Comment by fezproof** [comment]
I am going to give this a crack as just a library at first, but hopefully it will be easy to port over to kit itself later down the line. At the moment I am going to work with some assumption on acceptance criteria to make things a little simpler:

Adding a Form component, not an action
 - Will act in a similar way to [Remix's Form component](https://remix.run/docs/en/v1.5.1/api/remix#form) (If it ain't broke...)
 - Can use https://kit.svelte.dev/docs/configuration#methodoverride internally
 - Will assume a default action url if none is provided (same as current URL, similar to https://kit.svelte.dev/docs/routing#endpoints-page-endpoints)
 - Will automatically invalidate action URL on successful submit

## Interface
Current thoughts on interface, I think event dispatchers  and component bindings could be leveraged a bit here to be explicit about what the form is actually doing. It is also much less verbose

```svelte
<script>
  let state;
  let result;
  let data;
</script>

<!-- method can be defaulted to POST and action to page endpoint -->
<Form method="patch" action="/user" bind:state bind:data bind:result>
  <input type="text" name="age" />
  {#if result.error}
    <p>{result.error.message}</p>
  {/#if}
  <button type="submit" disabled={state === "submitting"}>Submit</button>
</Form>
```

### state:
`idle`: Form isn't currently doing anything
`submitting`: Form is submitting data to the server
`invalidating`: Form is invalidating endpoints after successful submission

### data:
The current form data
This may need to be function for performance reasons
Optimistic rendering could be achieved by using a combination of the forms state and data 

### result:
Result of submitting the form. This value is undefined unless successfully submitted

## Things to work out
 - Should the component swallow errors? (probably not)
   - I think we should point developers in the direction of explicit error handling on the server side
   - Perhaps some default errors could exist for common use cases? (Network errors)
 - Should the form invalidate just the
   - action endpoint?
   - layout endpoints?
   - an array that can be passed as a prop?
 - Should the form create new entries in the history stack or not?
 - Styling...
   - Probably could just pass in classes or something, this isn't the best
 - Probably a bunch of other stuff

## CSRF protection
I feel this is a much more complicated issue, and a robust method should be chosen outside of this issue. I will attempt this after the first set of criteria is met.

Does this kind of style make sense, or am I making some huge mistakes 😅

**Comment by fezproof** [comment]
Was playing around with this a bit, and I feel the component is the way to go, as it allows you to use slot props!
I mocked up a quick implementation and it allows you to do some cool stuff like this:

```svelte
<Form
  action="/todos"
  method="post"
  let:state
  on:complete={({ detail }) => {
    detail.form.reset();
  }}
>
  <input
    class:submitting={state !== 'idle'}
    name="text"
    aria-label="Add todo"
    placeholder="+ tap to add a todo"
  />
</Form>
```

The lack of classing is not really an issue to me, it just generally results in an extra div 🤷

[Here is a basic implementation](https://github.com/fezproof/sveltekit-form/blob/main/src/lib/Form.svelte) I did in an example project (I am still ignoring a lot of details, including CSRF). It works surprisingly well for the basic use case. I will be improving it and testing more scenarios over time. I will post more findings if I have any.

**Comment by fezproof** [comment]
Something I have noticed while trying to improve my implementation is that invalidate does not auto cancel duplicate calls. This can lead to very confusing and inconsistent behaviour. 

It can be even seen now in the current TODOs example when deleting a bunch of records at once. I even accidentally deleted a todo once due to the the page shifting because of this.

Should invalidate cancel duplicate requests automagically?

**Comment by NoelJacob** [comment]
I chose using \<form use:xxx\> over \<svelte:form\> over \<Form\>. Because it would still work with css libs which apply on \<form\>. 

Also it would be nice if \<svelte:form\> and other \<svelte:xxx\> got the css applied to \<xxx\> and \<form\>. It might seem like this css selection might confuse developers but, if they are using the \<svelte:xxx\>, it means they've read docs or come across the feature online, where they'll _probably_ be warned of this. Just a suggestion.

**Comment by Mlocik97** [comment]
Can we for now expose existing enchance action, so instead of having it in `$lib/` we could be able to import it from `@sveltejs/kit/enchance` or something like that?

**Comment by studylet** [comment]
Is there any way to use await block when the request is pending with use:enhance when it is client-side rendering?

For example:
```
<script>
   export let data;
<script/>
{#await data} 
   some flashing cards visual 
{:then data} 
   cards filled with data 
{/await}
```

**Comment by ulvidamirli** [comment]
Is there any way to use await block when the request is pending with use:enhance when it is client-side rendering?

For example:
```
<script>
   export let data;
<script/>

{#await data} 
   some flashing cards visual 
{:then data} 
   cards filled with data 
{/await}
```

**Comment by abdo643-HULK** [comment]
> Is there any way to use await block when the request is pending with use:enhance when it is client-side rendering?
> 
> For example:
> 
> ```
> <script>
>    export let data;
> <script/>
> 
> {#await data} 
>    some flashing cards visual 
> {:then data} 
>    cards filled with data 
> {/await}
> ```

You can do that by creating a promise in the pending function and resolving that Promise in the result. I made a example that uses a similar Pattern. 
https://svelte.dev/repl/26fc3f6e57ad4223ae53f95b30fe4f34?version=3.49.0

**Comment by bato3** [comment]
### CSRF
Instead one complex solution is better provide small parts, with the help of which users can compose the desired solution. 
[My proposal](https://github.com/sveltejs/kit/issues/72#issuecomment-1213619512)

OT: URL for CSRF in first post is broken (remove `3533#` from it)

**Comment by Mlocik97** [comment]
@bato3 would it work without JS?

**Comment by lukaszpolowczyk** [comment]
I must note my suggestion Targeted Slots - https://github.com/sveltejs/rfcs/pull/68
Targeted Slots would be ideal for the problem in this thread.

I previously cited Declarative Actions, but the author agreed that Targeted Slots solves his problem completely, solving other problems as well.

```svelte
<script>
  import { Form } from '$app/navigation';
</script>
<Form>
  <svelte:element slot="form"
    action="/api/test?_method=PUT" // manually method overrides
    method="post"
    class="flex flex-column p-4 mt-2"
    id="nativeFormELement"
    on:pending={(e) => console.log(e.detail /* => data */)}
    on:result={(e) => console.log(e.detail /* => response */)}
    on:error={(e) => console.log(e.detail /* => Error handler */)}
  >
    <input name="description" aria-label="Add todo" placeholder="+ tap to add a todo">
  </svelte:element>
</Form>
```

On the surface it looks usual, the magic of the proposal happens inside - in `Form` with Targeted Slots syntax.  
Syntax description - https://github.com/sveltejs/rfcs/pull/68
